<a href="https://colab.research.google.com/github/Narendraiitg/ANN_MultiClass_Classifier_on_Iris/blob/main/ANN_MultiClass_Classifier_on_Iris.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from google.colab import files

uploaded = files.upload()

for fn in uploaded.keys():
  print('User uploaded file "{name}" with length {length} bytes'.format(
      name=fn, length=len(uploaded[fn])))

Saving iris.data to iris.data
User uploaded file "iris.data" with length 4551 bytes


In [2]:
import pandas as pd
import io 
data= pd.read_csv(io.StringIO(uploaded['iris.data'].decode('utf-8')))
data.head()

5.1  3.5  1.4  0.2  Iris-setosa
0  4.9  3.0  1.4  0.2  Iris-setosa
1  4.7  3.2  1.3  0.2  Iris-setosa
2  4.6  3.1  1.5  0.2  Iris-setosa
3  5.0  3.6  1.4  0.2  Iris-setosa
4  5.4  3.9  1.7  0.4  Iris-setosa

In [4]:
# Importing the various libraries
import seaborn as sns
import numpy as np
import keras
from keras.models import Sequential
from keras.optimizers import Adam
from keras.utils.vis_utils import model_to_dot
from keras.utils import plot_model
import pydot
import graphviz
from IPython.display import SVG

In [5]:
dataset= pd.read_csv("iris.data")
dataset

5.1  3.5  1.4  0.2     Iris-setosa
0    4.9  3.0  1.4  0.2     Iris-setosa
1    4.7  3.2  1.3  0.2     Iris-setosa
2    4.6  3.1  1.5  0.2     Iris-setosa
3    5.0  3.6  1.4  0.2     Iris-setosa
4    5.4  3.9  1.7  0.4     Iris-setosa
..   ...  ...  ...  ...             ...
144  6.7  3.0  5.2  2.3  Iris-virginica
145  6.3  2.5  5.0  1.9  Iris-virginica
146  6.5  3.0  5.2  2.0  Iris-virginica
147  6.2  3.4  5.4  2.3  Iris-virginica
148  5.9  3.0  5.1  1.8  Iris-virginica

[149 rows x 5 columns]

# Data Preprocessing

In [6]:
# Seperating the dependent and independent variables
X= data.iloc[:,0:4].values
y= data.iloc[:, 4].values

In [9]:
# Encoding categorial data 
from sklearn.preprocessing import LabelEncoder
LE= LabelEncoder()
y1= LE.fit_transform(y)
Y= pd.get_dummies(y1).values # converting the different values of y1 in the form of matrix

In [10]:
y1

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [11]:
Y

array([[1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [1, 0, 0],
       [0, 1, 0],
       [0, 1, 0],
       [0, 1, 0],
       [0, 1, 0],
       [0, 1, 0],
       [0, 1, 0],
       [0,

In above [1,0,0] represent Iris-setosa, [0,1,0] represent Iris-versicolor and [0,0,1] represent Iris-virginica

In [36]:
# Spiliting the data set into training and testing 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test= train_test_split(X,Y, test_size=0.3, random_state=0)


In [37]:
from keras.models import Sequential
from keras.layers import Dense

In [38]:
# Adding the input and first hidden layer 
model= Sequential()

#The first hidden layer has 16 nodes and uses the relu activation function.
model.add(Dense(4, input_shape=(4,), activation='relu'))

# Adding the output layer
model.add(Dense(3, activation='softmax')) # here we used softmax activation in output which will decide the output on the basis of max prob of all 3 catg of o/p


In [39]:
model.compile(optimizer="Adam", loss='categorical_crossentropy', metrics= ['accuracy'])

In [43]:
model.fit(X_train, y_train, batch_size=10, epochs=100, verbose=0)

In [44]:
# make predictions with the model
y_pred = model.predict(X_test)
y_test_class= np.argmax(y_test,axis=1)
y_pred_class= np.argmax(y_pred,axis=1)

2/2 [==============================] - 0s 5ms/step


In [45]:
# Making the confusion matrix
from sklearn.metrics import classification_report, confusion_matrix
print(classification_report(y_test_class, y_pred_class))
print(confusion_matrix(y_test_class, y_pred_class))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       1.00      0.94      0.97        16
           2       0.93      1.00      0.96        13

    accuracy                           0.98        45
   macro avg       0.98      0.98      0.98        45
weighted avg       0.98      0.98      0.98        45

[[16  0  0]
 [ 0 15  1]
 [ 0  0 13]]


From above confusion matrix acuuracy is sum of diagonal elements devide by total sum of elements = (16+15+13)/45 = 0.98 